### PR TITLE
[Build][Test] Add logic for image shift for AutoPatch Images

### PR DESCRIPTION
**Pull request by shantanutrip**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
This PR adds the logic to transfer the images from AutoPatch ECR to Beta ECR.

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/image_transfer.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    :param autopatch_image_uri: str, Image URI of the AutoPatch image
+    :param beta_image_uri: str, Image URI of the Beta image
+    :param image_transfer_override_flags: dict, Image Override flags derived from the s3 Bucket
+    :return: boolean, True if the transfer can happen, False otherwise.
+    """
+    if is_image_transfer_enabled_by_override_flags(
+        image_uri=autopatch_image_uri, image_transfer_override_flags=image_transfer_override_flags
+    ):
+        return True
+    beta_image_push_time = get_push_time_of_image_from_ecr(image_uri=beta_image_uri)
```
Comment (saimidu) [inline src/image_transfer.py:190]: If this script already imports `test.test_utils`, then why does it need the use of a wrapper function around `get_image_push_time_from_ecr` in test_utils?

File: src/image_transfer.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+def get_push_time_of_image_from_ecr(image_uri):
+    """
+    This method uses the ERC boto3 client to get the push time of an image.
+
+    :param image_uri: str, Image URI
+    :return: datetime.datetime Object, Time of the Push
+    """
+    image_region = test_utils.get_region_from_image_uri(image_uri=image_uri)
+    ecr_client = boto3.client("ecr", region_name=image_region)
+    return test_utils.get_image_push_time_from_ecr(ecr_client=ecr_client, image_uri=image_uri)
```
Comment (saimidu) [inline src/image_transfer.py:70]: Why not just use this function instead of creating a wrapper for it here?

File: src/image_transfer.py (hunk 1 above)
Comment (shantanutrip) [reply src/image_transfer.py]: That's a nice suggestion. I have removed the wrapper function altogether. Thanks!

File: src/image_transfer.py (hunk 2 above)
Comment (shantanutrip) [reply src/image_transfer.py]: That's a nice suggestion. I have removed the wrapper function altogether. Thanks!

File: src/image_transfer.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    otherwise False.
+
+    :param image_uri: str, Image URI
+    :param image_transfer_override_flags: dict, Image Override flags derived from the s3 Bucket
+    """
+    commit_id = os.getenv("CODEBUILD_RESOLVED_SOURCE_VERSION", "default")
+    if commit_id in image_transfer_override_flags:
+        if (
+            not image_transfer_override_flags[commit_id]
+            or image_uri in image_transfer_override_flags[commit_id]
```
[the remark this answers is not in the document]
Comment (shantanutrip) [reply src/image_transfer.py]: Great catch! I have changed the condition now. 

The `image_uri` is extracted using this method: https://github.com/aws/deep-learning-containers/blob/master/test/test_utils/__init__.py#L1124 that is also used by our tests. One would just need to look into the logs of the CodeBuild job of the Build Stage to get the image uris.

File: src/image_transfer.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        # If Beta image was built more than 5 days ago
+        return True
+    elif is_latest_benchmark_tested_beta_image_an_autopatch_image_itself(
+        beta_image_uri=beta_image_uri
+    ):
+        # If the latest benchmark tested image is autopatch image itself, transfer the image
+        LOGGER.info(
+            f"{autopatch_image_uri} is transferable since {beta_image_uri} is autopatch image itself."
+        )
+        return True
```
[the remark this answers is not in the document]
Comment (shantanutrip) [reply src/image_transfer.py]: I have removed this logic and added a `TODO` statement for future addition.

**Review by saimidu — APPROVED** [review]

**Review by saimidu — APPROVED** [review]

**Pull request merged**